p2p/transport/webrtc/udpmux/muxed_connection.go

```diff
@@ -0,0 +1,178 @@
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	pool "github.com/libp2p/go-buffer-pool"
+)
+
+var _ net.PacketConn = &muxedConnection{}
+
+const maxPacketsInQueue int = 128
+
+// muxedConnection provides a net.PacketConn abstraction
+// over packetQueue and adds the ability to store addresses
+// from which this connection (indexed by ufrag) received
+// data.
+type muxedConnection struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	pq     *packetQueue
+	// list of remote addresses associated with this connection.
+	// this is useful as a mapping from [address] -> ufrag
+	addresses []string
+	ufrag     string
+	mux       *udpMux
+}
+
+func newMuxedConnection(mux *udpMux, ufrag string) *muxedConnection {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &muxedConnection{
+		ctx:    ctx,
+		cancel: cancel,
+		pq:     newPacketQueue(),
+		ufrag:  ufrag,
+		mux:    mux,
+	}
+}
+
+func (conn *muxedConnection) push(buf []byte, addr net.Addr) error {
+	return conn.pq.push(buf, addr)
+}
+
+// Close implements net.PacketConn
+func (conn *muxedConnection) Close() error {
+	_ = conn.closeConnection()
+	conn.mux.RemoveConnByUfrag(conn.ufrag)
+	return nil
+}
+
+// LocalAddr implements net.PacketConn
+func (conn *muxedConnection) LocalAddr() net.Addr {
+	return conn.mux.socket.LocalAddr()
+}
+
+// ReadFrom implements net.PacketConn
+func (conn *muxedConnection) ReadFrom(p []byte) (n int, addr net.Addr, err error) {
+	return conn.pq.pop(conn.ctx, p)
+}
+
+// SetDeadline implements net.PacketConn
+func (*muxedConnection) SetDeadline(t time.Time) error {
+	return nil
+}
+
+// SetReadDeadline implements net.PacketConn
+func (*muxedConnection) SetReadDeadline(t time.Time) error {
+	return nil
+}
+
+// SetWriteDeadline implements net.PacketConn
+func (*muxedConnection) SetWriteDeadline(t time.Time) error {
+	return nil
+}
+
+// WriteTo implements net.PacketConn
+func (conn *muxedConnection) WriteTo(p []byte, addr net.Addr) (n int, err error) {
+	return conn.mux.writeTo(p, addr)
+}
+
+func (conn *muxedConnection) closeConnection() error {
+	select {
+	case <-conn.ctx.Done():
+		return fmt.Errorf("already closed")
+	default:
+	}
+	conn.pq.close()
+	conn.cancel()
+	return nil
+}
+
+type packet struct {
+	addr net.Addr
+	buf  []byte
+}
+
+var (
+	errTooManyPackets = fmt.Errorf("too many packets in queue; dropping")
+)
+
+// just a convenience wrapper around a channel
+type packetQueue struct {
+	sync.Mutex
+	pkts   chan packet
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func newPacketQueue() *packetQueue {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &packetQueue{
+		ctx:    ctx,
+		cancel: cancel,
+		pkts:   make(chan packet, maxPacketsInQueue),
+	}
+}
+
+// pop reads a packet from the packetQueue or blocks until
+// either a packet becomes available or the buffer is closed.
+func (pq *packetQueue) pop(ctx context.Context, buf []byte) (int, net.Addr, error) {
+	select {
```

Comment:
TODO: cannot put my finger on it yet, but this code smells,
it seems like you are mixing logic boundaries here,
making it harder to reason about the code

Reply:
The reason this does not return when the internal context's `Done()` returns is because there could still be packets in the channel that could be read even if the channel is closed. It should ideally be a priority select with priority for reading from the channel, followed by checking if the context is closed.

Reply:
fixed in https://github.com/libp2p/go-libp2p/pull/1999/commits/f76dfc499857bffe287bedd931eeba6c652a4825 (first commit of new PR https://github.com/libp2p/go-libp2p/pull/1999)